newIDE/app/src/VersionHistory/UseVersionHistory.js

```diff
@@ -0,0 +1,441 @@
+// @flow
+
+import * as React from 'react';
+import { t, Trans } from '@lingui/macro';
+import Drawer from '@material-ui/core/Drawer';
+import DrawerTopBar from '../UI/DrawerTopBar';
+import {
+  listVersionsOfProject,
+  type ExpandedCloudProjectVersion,
+  updateCloudProjectVersion,
+} from '../Utils/GDevelopServices/Project';
+import type { FileMetadata, StorageProvider } from '../ProjectsStorage';
+import AuthenticatedUserContext from '../Profile/AuthenticatedUserContext';
+import { canUseCloudProjectHistory } from '../Utils/GDevelopServices/Usage';
+import { Column, Line } from '../UI/Grid';
+import VersionHistory, { type OpenedVersionStatus } from '.';
+import UnsavedChangesContext from '../MainFrame/UnsavedChangesContext';
+import AlertMessage from '../UI/AlertMessage';
+import { ColumnStackLayout } from '../UI/Layout';
+import RaisedButton from '../UI/RaisedButton';
+import { SubscriptionSuggestionContext } from '../Profile/Subscription/SubscriptionSuggestionContext';
+import useAlertDialog from '../UI/Alert/useAlertDialog';
+import PlaceholderLoader from '../UI/PlaceholderLoader';
+import type { MessageDescriptor } from '../Utils/i18n/MessageDescriptor.flow';
+
+const styles = {
+  drawerContent: {
+    width: 320,
+    overflowX: 'hidden',
+    display: 'flex',
+    flexDirection: 'column',
+  },
+};
+
+const mergeVersionsLists = (
+  list1: ExpandedCloudProjectVersion[],
+  list2: ExpandedCloudProjectVersion[]
+) => {
+  const mostRecentVersionDateInList2 = Date.parse(list2[0].createdAt);
+  const moreRecentVersionsInList1 = list1.filter(
+    version => Date.parse(version.createdAt) > mostRecentVersionDateInList2
+  );
+  return [...moreRecentVersionsInList1, ...list2];
+};
+
+type PaginationState = {|
+  versions: ?(ExpandedCloudProjectVersion[]),
+  nextPageUri: ?Object,
+|};
+
+const emptyPaginationState: PaginationState = {
+  versions: null,
+  nextPageUri: null,
+};
+
+type Props = {|
+  getStorageProvider: () => StorageProvider,
+  fileMetadata: ?FileMetadata,
+  isSavingProject: boolean,
+  onOpenCloudProjectOnSpecificVersion: ({|
+    fileMetadata: FileMetadata,
+    versionId: string,
+    ignoreUnsavedChanges: boolean,
+    openingMessage: MessageDescriptor,
+  |}) => Promise<void>,
+|};
+
+const useVersionHistory = ({
+  fileMetadata,
+  isSavingProject,
+  getStorageProvider,
+  onOpenCloudProjectOnSpecificVersion,
+}: Props) => {
+  const { hasUnsavedChanges } = React.useContext(UnsavedChangesContext);
+  const { showAlert } = useAlertDialog();
+  const { openSubscriptionDialog } = React.useContext(
+    SubscriptionSuggestionContext
+  );
+  const authenticatedUser = React.useContext(AuthenticatedUserContext);
+  const ignoreFileMetadataChangesRef = React.useRef<boolean>(false);
+  const preventEffectsRunningRef = React.useRef<boolean>(false);
```

Comment:
Any chance this could have a name explaining its usage, a bit like "ignoreFileMetadataChanges"? "freezeWhileLoadingSpecificVersionRef"?